src/Microsoft.Diagnostics.DebugServices/MemoryService.cs

```diff
@@ -0,0 +1,290 @@
+        /// </summary>
+        /// <param name="dataReader">CLRMD data reader</param>
+        public MemoryService(IDataReader dataReader)
+        {
+            _dataReader = dataReader;
+            _memoryCache = new MemoryCache(ReadMemoryFromModule);
+        }
+
+        /// <summary>
+        /// Read memory out of the target process.
+        /// </summary>
+        /// <param name="address">The address of memory to read</param>
+        /// <param name="buffer">The buffer to write to</param>
+        /// <param name="bytesRequested">The number of bytes to read</param>
+        /// <param name="bytesRead">The number of bytes actually read out of the target process</param>
+        /// <returns>true if any bytes were read at all, false if the read failed (and no bytes were read)</returns>
+        public bool ReadMemory(ulong address, byte[] buffer, int bytesRequested, out int bytesRead)
+        {
+            return ReadMemory(address, new Span<byte>(buffer, 0, bytesRequested), out bytesRead);
+        }
+
+        /// <summary>
+        /// Read memory out of the target process.
+        /// </summary>
+        /// <param name="address">The address of memory to read</param>
+        /// <param name="buffer">The buffer to read memory into</param>
+        /// <param name="bytesRequested">The number of bytes to read</param>
+        /// <param name="bytesRead">The number of bytes actually read out of the target process</param>
+        /// <returns>true if any bytes were read at all, false if the read failed (and no bytes were read)</returns>
+        public bool ReadMemory(ulong address, IntPtr buffer, int bytesRequested, out int bytesRead)
+        {
+            unsafe
+            {
+                return ReadMemory(address, new Span<byte>(buffer.ToPointer(), bytesRequested), out bytesRead);
+            }
+        }
+
+        /// <summary>
+        /// Read memory out of the target process.
+        /// </summary>
+        /// <param name="address">The address of memory to read</param>
+        /// <param name="buffer">The buffer to read memory into</param>
+        /// <param name="bytesRead">The number of bytes actually read out of the target process</param>
+        /// <returns>true if any bytes were read at all, false if the read failed (and no bytes were read)</returns>
+        public bool ReadMemory(ulong address, Span<byte> buffer, out int bytesRead)
+        {
+            int bytesRequested = buffer.Length;
+            bool result = false;
+            unsafe
+            {
+                fixed (byte* ptr = buffer)
+                {
+                    result = _dataReader.ReadMemory(address, new IntPtr(ptr), bytesRequested, out bytesRead);
+                }
+            }
+            // If the read failed or a successful partial read
+            if (!result || (bytesRequested != bytesRead))
+            {
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```

Comment:
If we are doing all the work to map images on windows but not on other OSes, does that mean the other OSes are now less functional or there is a reason why the issues don't repro there?

Reply:
The "new" dumpheap testing found problems in the Windows minidump read memory that "maps" the actual PE module into the virtual space. The new tests caused the DAC/SOS to read sections of the PE modules that had relocations but were not fixed up in the original module mapping code. This only happens on Windows minidumps because Linux core dumps have all the necessary fixed up sections (except the metadata which is handled in the native SOS code for xplat) and Windows minidumps don't.